README.md
# evoBox
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://choosealicense.com/licenses/mit/l)

Genetic evolution / creature simulator.

CHANGELOG.md
# evoBox Changelog

## [DEV] v2.0.0 (April 18th)

### Features
- Increased default UPS aim to 50 (10 > 50)
- Rebalanced all rates to be not dependent on UPS
- Rebalanced and tweaked some rates
- Added cause of death to log when creatures dies
- Added UPS to text in top left
- Added emoji mode (again)
- Made target lines dashed
- Removed frame rate aim / lock
- Removed background animation for performance

### Changes
- Reproduction can now only be with the opposite sex
- Lowered top left text


## v1.0.1 (March 27th)

### Features
- Added padding to avoid objects being against the edges
- Added occlusion culling (hiding unseen objects) to objects

### Changes
- Force renderer to update when zoom changes

### Fixes
- Fixed that sometimes genes were missed in the child after mating


## v1.0.0-RC2 (March 27th)

### Changes
- Upped frame rate lock to 60 (from 30)
- Decreased total background cells to 300 (from 500)


## v1.0.0 (March 26th)

### Features
- **Added diseases**
- Pressing 'r' now restarts the world
- Zooming with shortcuts Ctrl + key now zooms the sandbox instead of the actual page
- **Added coloring to log messages**
- Can now pan around by scrolling
- **New renderer using canvas, a lot more performant and uses less resources**
- Updated panning border restriction to be smoother
- **Added new random multicolor moving background**
- Updated some styling of the UI

### Changes
- Auto exit full zoom mode on attempt to change zoom
- Unmade dead creatures degrade
- Made controls window's width is smaller

### Fixes
- Fixed creatures able to mate with dead creatures
- Fixed children not having genes after mating
- Fixed children from asexual reproduction's genes being linked to their parent (if either change, they both change)


## v0.5.0 (March 19th)

### Features
- Double click anywhere in sandbox to add food
- Added controls window
- Added fadein and fadeout animations to windows
- Food and age of dead creatures continue to change
- Added dead attribute to creature window

### Changes
- Made windows more transparent

### Fixes
- Fixed windows being able to be dragged from their body, only header


## v0.4.0 (March 15th)

### Features
- Improved creature name generation
- Last names now stay with children
- Added border to sandbox
- Smooth entity movement
- Added mode to have sandbox width = screen width
- Creatures now have random colo(u)rs
- Creatures size is now scaling with their mass / food
- Added watermark
- Added reset / restart button
- Made top bar larger on larger displays
- Added new logo

### Changes
- Removed version from title
- Food is now green instead of red
- Updated meta / description tags

### Fixes
- Force UI updates on entities when zooming (fixes nothing happening when paused and changing zoom)
- Disable pointer cursor for food
- Fixed creature removal event reporting the 'removeFood' event instead of the 'removeCreature' event

## v0.3.0 (March 12th)

### Features
- Add to log when creature reproduces
- Added zooming, panning and relative space
- Added Unicode / emojis to things in top bar
- Added partices for mutating, reproducing and dying
- Added window popup when clicking creatures including information

### Fixes
- Fixed some symbol fonts incorrectly rendering the time symbols
- Readded age death for creatures


## v0.2.0 (March 9th)

### Features
- Enhanced creature's food choices
- When there's more than one line in the log, now fade out the older lines instead of just removing them

### Changes
- Decreased natural food quality degrade rate
- Moved dead checks to end of updates (creatures + food)
- Stopped creature's not going to food if another creature is targetting it as all creature's targets would be unknown to a single creature

### Fixes
- Fixed that food would not run update callbacks
- Fixed food having negative values and not being destroyed
- Fixed food value still working with negative quality
- Fixed creature eating being inaccurate and unstable
- Fixed food spawning rates being inverse for speeds (fast speed = slower rate, slow speed = faster rate)
- Fixed natural food quality degrade using old speed factor function


## v0.1.0 (March 8th)
Initial release.

src/js/ui/main.js
import * as UICreature from './creature';
import * as UIFood from './food';
import * as UIUtil from './util';
import SimFood from '/js/sim/food';

import { Window, dragElement } from './window';

import * as Renderer from './renderer';

let screenSize = 'full';

async function creatureCallback(type, data) {
  let f = UICreature[type];

  if (f !== undefined) {
    f(data);
  }
}

async function foodCallback(type, data) {
  let f = UIFood[type];

  if (f !== undefined) {
    f(data);
  }
}

async function worldCallback(type, data) {
  if (type === 'addCreature') {
    data.eventCallback = creatureCallback;
    //UICreature.create(data);
  }

  if (type === 'removeCreature') {
    //UICreature.destroy(data);
  }

  if (type === 'addFood') {
    data.eventCallback = foodCallback;
    //UIFood.create(data);
  }

  if (type === 'removeFood') {
    //UIFood.destroy(data);
  }

  if (type === 'update') {
    //Renderer.update(data);

    document.getElementById('play').style.display = data.paused ? 'inline' : 'none';
    document.getElementById('pause').style.display = data.paused ? 'none' : 'inline';

    document.getElementById('sf').innerText = `${UIUtil.showNumber(data.speed)}x`;
    document.getElementById('zf').innerText = `${UIUtil.showNumber(UIUtil.sizeFactor())}x`;

    document.getElementById('elapsed').innerText = `${elapsedTime(data)}`;

    if (screenSize === 'full') {
      let sandbox = document.getElementById('sandbox');
      let cWidth = sandbox.style.width.replace('px', '');
      let sWidth = window.innerWidth;
      if (window.innerHeight > sWidth) {
        sWidth = window.innerHeight;
      }

      // let diff = cWidth - sWidth;

      if (cWidth != sWidth) {
        sandbox.style.width = `${sWidth}px`;
        sandbox.style.height = `${sWidth}px`;
        sandbox.width = sWidth;
        sandbox.height = sWidth;

        let top = 0; //parseFloat(sandbox.style.top.replace('px', '')) + diff;
        top = top > 0 ? 0 : top;

        sandbox.style.top = `${top}px`;

        let left = 0; //parseFloat(sandbox.style.left.replace('px', '')) + diff;
        left = left > 0 ? 0 : left;

        sandbox.style.left = `${left}px`;
      }
    }
  }
}

function elapsedTime(world) {
  let secs = Math.floor(world.elapsedTime / 1000);

  if (secs >= 60) {
    let mins = Math.floor(secs / 60);
    secs = Math.floor(secs % 60);

    secs = secs.toString().length === 1 ? `0${secs}` : secs;

    return `${mins}:${secs}`
  }

  return secs.toString();
}

export function UIWorld(world) {
  world.eventCallback = worldCallback;

  window.onload = function() {
    Renderer.init(world);

    UIInit(world);
  };

  world.options.emojiMode = false;
}


function increaseSpeed(world) {
  world.speed = world.speed >= 5 ? world.speed : world.speed + 0.25;
}

function decreaseSpeed(world) {
  world.speed = world.speed <= 0.25 ? world.speed : world.speed - 0.25;
}

function showControls() {
  let win = new Window('Controls', 10, 10);
  win[0].style.height = '410px';
  win[0].style.width = '200px';

  let rows = [['Speed Up', '>'], ['Slow Down', '<'], ['Pause', 'Space'], ['', ''], ['Zoom Out', '-'], ['Zoom In', '='], ['Full Mode', 'f'], ['', ''], ['Restart World', 'r'], ['Controls', '?'], ['Debug', 'd'], ['🙂 🇲​🇴​🇩​🇪', 'e']];

  for (let r of rows) {
    let row = document.createElement('div');
    row.className = 'row';

    let rowLeft = document.createElement('div');
    rowLeft.className = 'row-left';
    rowLeft.innerText = r[0];

    let rowRight = document.createElement('div');
    rowRight.className = 'row-right';
    rowRight.innerText = r[1];

    row.appendChild(rowLeft);
    row.appendChild(rowRight);

    win[1].appendChild(row);
  }
}

function zoomIn(world) {
  let sandbox = document.getElementById('sandbox');
  let width = sandbox.style.width ? sandbox.style.width : '2000px';
  let size = parseFloat(width.replace('px', '')) + 500;

  sandbox.style.width = `${size}px`;
  sandbox.style.height = `${size}px`;
  sandbox.width = size;
  sandbox.height = size;

  screenSize = 'dynamic';

  Renderer.update(world);
}

function zoomOut(world) {
  let sandbox = document.getElementById('sandbox');
  let width = sandbox.style.width ? sandbox.style.width : '2000px';
  let size = parseFloat(width.replace('px', '')) - 500;
  size = size < window.innerWidth ? (Math.floor(window.innerWidth / 500) + 1) * 500 : size;

  let changed = size !== width;

  sandbox.style.width = `${size}px`;
  sandbox.style.height = `${size}px`;
  sandbox.width = size;
  sandbox.height = size;

  if (changed) {
    let top = parseFloat(sandbox.style.top.replace('px', '')) + 500;
    top = top > 0 ? 0 : top;

    sandbox.style.top = `${top}px`;

    let left = parseFloat(sandbox.style.left.replace('px', '')) + 500;
    left = left > 0 ? 0 : left;

    sandbox.style.left = `${left}px`;
  }

  screenSize = 'dynamic';

  Renderer.update(world);
}

let lastKey = undefined;

export function UIInit(world) {
  let fired = false;

  showControls();

  document.onkeypress = function (e) {
    if (!fired) {
      console.log(e);

      if (e.code === 'Space') {
        world.paused = !world.paused;
      }

      if (e.key === '>') {
        increaseSpeed(world);
      }

      if (e.key === '<') {
        decreaseSpeed(world);
      }

      if (e.key === 'd') {
        let debug = document.getElementById('debug');
        document.getElementById('debug').className = debug.className === '' ? 'show' : '';
      }

      if (e.key === '=') {
        zoomIn(world);
      }

      if (e.key === '-') {
        zoomOut(world);
      }

      if (e.key === 'f') {
        screenSize = screenSize === 'full' ? 'dynamic' : 'full';
      }

      if (e.key === 'e') {
        world.options.emojiMode = !world.options.emojiMode;
      }

      if (e.key === 'r') {
        world.restart();
      }

      if (e.key === '?') {
        showControls();
      }

      fired = true;
      lastKey = e.key;
    }

    e.preventDefault();
  };

  document.onkeyup = function() {
    fired = false;
  };

  document.onkeydown = function(event) {
    if (event.ctrlKey === true) {
      if (event.key === '-') {
        zoomOut(world);

        event.preventDefault();
      }

      if (event.key === '=') {
        zoomIn(world);

        event.preventDefault();
      }
    }  
  };

  document.onwheel = function(e) {
    let elmnt = document.getElementById('sandbox');

    let newTop = elmnt.offsetTop - e.deltaY;
    let newLeft = elmnt.offsetLeft - e.deltaX;

    newTop = newTop > 0 ? 0 : newTop;
    newLeft = newLeft > 0 ? 0 : newLeft;

    let adjTop = newTop + UIUtil.sizeAmount()
    if (adjTop < window.innerHeight) {
      newTop -= adjTop - window.innerHeight;
    }

    let adjLeft = newLeft + UIUtil.sizeAmount();
    if (adjLeft < window.innerWidth) {
      newLeft -= adjLeft - window.innerWidth;
    }

    //while (newTop + UIUtil.sizeAmount() < window.innerHeight) { newTop++; }

    //while (newLeft + UIUtil.sizeAmount() < window.innerWidth) { newLeft++; }

    elmnt.style.top = newTop + "px";
    elmnt.style.left = newLeft + "px";
  };

  document.getElementById('play').onclick = function() {
    world.paused = false;
  };

  document.getElementById('pause').onclick = function() {
    world.paused = true;
  };

  document.getElementById('slow').onclick = function() {
    decreaseSpeed(world);
  };

  document.getElementById('fast').onclick = function() {
    increaseSpeed(world);
  };

  document.getElementById('restart').onclick = function() {
    world.restart();
  };

  document.getElementById('sandbox').ondblclick = function(e) {
    let x = e.clientX - parseFloat(document.getElementById('sandbox').style.left.replace('px', ''));
    let y = e.clientY - parseFloat(document.getElementById('sandbox').style.top.replace('px', ''));

    let pos = UIUtil.getRelativePosition(x, y);

    world.addFood(new SimFood(world, undefined, undefined, pos.x, pos.y));
  };

  dragElement(document.getElementById('sandbox'));
}